sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeFormatterHelper.scala

```diff
@@ -104,8 +111,40 @@ private object DateTimeFormatterHelper {
       .withResolverStyle(ResolverStyle.STRICT)
   }
 
-  def buildFormatter(pattern: String, locale: Locale): DateTimeFormatter = {
-    val builder = createBuilder().appendPattern(pattern)
+  /**
+   * Building a formatter for parsing seconds fraction with variable length
+   */
+  def createBuilderWithVarLengthSecondFraction(
```

Comment:
How about move this method into `DateTimeUtils`, together with `convertIncompatiblePattern`?
I think we need to collect all these kind of util functions analyzing pattern string together.

Reply:
This is very specific to datetime parsing. How about we move `convertIncompatiblePattern` to `object DateTimeFormatterHelper`?

Reply:
Then, I will not do this refactoring in this PR, not related